pandas/core/dtypes/dtypes.py

```diff
@@ -110,37 +110,144 @@ class CategoricalDtypeType(type):
+        if categories is not None:
+            categories = Index(categories, tupleize_cols=False)
+            # validation
+            cls._validate_categories(categories, fastpath=fastpath)
+            cls._validate_ordered(ordered)
+        categorical = object.__new__(cls)
+        categorical._categories = categories
+        categorical._ordered = ordered
+        return categorical
 
     def __hash__(self):
-        # make myself hashable
-        return hash(str(self))
+        # _hash_categories returns a uint64, so use the negative
+        # space for when we have unknown categories to avoid a conflict
+        if self.categories is None:
+            if self.ordered:
+                return -1
+            else:
+                return -2
+        # We *do* want to include the real self.ordered here
+        return int(self._hash_categories(self.categories, self.ordered))
 
     def __eq__(self, other):
         if isinstance(other, compat.string_types):
             return other == self.name
 
-        return isinstance(other, CategoricalDtype)
+        if not (hasattr(other, 'ordered') and hasattr(other, 'categories')):
+            return False
+        elif self.categories is None or other.categories is None:
+            # We're forced into a suboptimal corner thanks to math and
+            # backwards compatibility. We require that `CDT(...) == 'category'`
+            # for all CDTs **including** `CDT(None, ...)`. Therefore, *all*
+            # CDT(., .) = CDT(None, False) and *all*
+            # CDT(., .) = CDT(None, True).
+            return True
+        elif self.ordered:
+            return other.ordered and self.categories.equals(other.categories)
+        elif other.ordered:
+            return False
+        else:
+            # both unordered; this could probably be optimized / cached
+            return hash(self) == hash(other)
+
+    def __unicode__(self):
+        tpl = u'CategoricalDtype({}ordered={})'
+        if self.categories is None:
+            data = u"None, "
+        else:
+            data = self.categories._format_data(name=self.__class__.__name__)
+        return tpl.format(data, self.ordered)
+
+    def __repr__(self):
+        return str(self)
+
+    def __getnewargs__(self):
+        return (self.categories, self.ordered)
+
+    @staticmethod
+    def _hash_categories(categories, ordered=True):
+        from pandas.core.util.hashing import (
+            hash_array, _combine_hash_arrays, hash_tuples
+        )
+
+        categories = np.asarray(categories)
+        if len(categories) and isinstance(categories[0], tuple):
+            # assumes if any individual category is a tuple, then all our. ATM
+            # I don't really want to support just some of the categories being
+            # tuples.
+            categories = list(categories)  # breaks if a np.array of categories
+            cat_array = hash_tuples(categories)
+        else:
+            if categories.dtype == 'O':
```

Comment:
we should actually move this to hash_array I think (could be pre-cursor PR)

Reply:
?

Reply:
Looking into this now.

Any thoughts on what this does to the probability of hash collisions? If we use two distinct hashing algorithms, I'd expect probability of a collision is higher (though presumably still negligible?),

Reply:
neh hash collisions basically don't happen in the space of things we are doing (normally). however you can collide by construction, e.g. code doing something wrong..

In any event we should move this code (can be a followup, maybe make an issue for things to do)